File: linearPharse.c
// origin https://gist.github.com/kmc7468/b07b82929030ca4a5ab8f77de67147d3 

/*
 * MIT License
 *
 * Copyright (c) 2021 kmc7468
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include <algorithm>
#include <cctype>
#include <cmath>
#include <cstddef>
#include <iostream>
#include <iterator>
#include <string>
#include <string_view>
#include <utility>

struct Equation {
	double Coefficient = 0;
	double Constant = 0;
};

bool ParseEquation(Equation& result, const std::string& equation);
double SolveEquation(Equation& equation);

struct ParseContext {
	std::string_view Equation;
	std::size_t Offset = 0;

	char Required = 0;
};

void SkipEmptyChars(ParseContext& context);
bool RequireChar(ParseContext& context, char c);
bool Error(std::size_t offset, const std::string_view& reason);

bool ParseAddSub(ParseContext& context, Equation& result);
bool ParseMulDiv(ParseContext& context, Equation& result);
bool ParseParen(ParseContext& context, Equation& result);
bool ParseTerm(ParseContext& context, Equation& result);

constexpr std::string_view g_InputMessage = "x에 대한 일차방정식을 입력하세요: ";

int main() {
	std::cout << "일차방정식 계산 프로그램\n"
		"(C) 2021. kmc7468 All rights reserved.\n"
		"주의사항: 괄호 앞/뒤에서 곱셈 연산자를 생략할 수 없습니다.\n\n";

startPoint:
	std::string equation;
	std::cout << g_InputMessage;
	std::getline(std::cin, equation);

	Equation result;
	if (!ParseEquation(result, equation)) goto startPoint;

	double root = SolveEquation(result);
	if (std::isinf(root)) {
		std::cout << "해가 무수히 많습니다.\n\n";
	} else if (std::isnan(root)) {
		std::cout << "해가 없습니다.\n\n";
	} else {
		std::cout << "해는 x=" << root << " 입니다.\n\n";
	}

	goto startPoint;
}

bool ParseEquation(Equation& result, const std::string& equation) {
	ParseContext context;
	context.Equation = equation;

	Equation lhs, rhs;
	if (!ParseAddSub(context, lhs)) return false;
	else if (!RequireChar(context, '=')) return Error(context.Offset, "등호가 없습니다.");
	else if (!ParseAddSub(context, rhs)) return false;
	else if (context.Offset < equation.size()) return Error(context.Offset, "올바른 형식의 일차방정식이 아닙니다.");

	result.Coefficient = lhs.Coefficient - rhs.Coefficient;
	result.Constant = rhs.Constant - lhs.Constant;

	return true;
}
double SolveEquation(Equation& equation) {
	if (equation.Coefficient == 0)
		if (equation.Constant == 0) return INFINITY;
		else return NAN;
	else return equation.Constant / equation.Coefficient;
}

void SkipEmptyChars(ParseContext& context) {
	while (context.Offset < context.Equation.size() && std::isspace(context.Equation[context.Offset])) {
		++context.Offset;
	}
}
bool RequireChar(ParseContext& context, char c) {
	SkipEmptyChars(context);

	if (context.Offset < context.Equation.size() && context.Equation[context.Offset] == c) {
		++context.Offset;
		context.Required = c;

		return true;
	} else return false;
}
bool Error(std::size_t offset, const std::string_view& reason) {
	std::cout << std::string(g_InputMessage.size() + offset, ' ') << "^\n";
	std::cout << "오류: " << reason << "\n\n";

	return false;
}

bool ParseAddSub(ParseContext& context, Equation& result) {
	if (!ParseMulDiv(context, result)) return false;

	while (RequireChar(context, '+') || RequireChar(context, '-')) {
		const double sign = context.Required == '+' ? 1 : -1;
		Equation rhs;
		if (!ParseMulDiv(context, rhs)) return false;

		result.Coefficient += sign * rhs.Coefficient;
		result.Constant += sign * rhs.Constant;
	}

	return true;
}
bool ParseMulDiv(ParseContext& context, Equation& result) {
	if (!ParseParen(context, result)) return false;

	while (RequireChar(context, '*') || RequireChar(context, '/')) {
		const double exp = context.Required == '*' ? 1 : -1;
		Equation rhs;
		if (!ParseParen(context, rhs)) return false;
		else if (exp == -1 && rhs.Coefficient == 0 && rhs.Constant == 0) return Error(context.Offset - 1, "0으로 나눌 수 없습니다.");

		if (exp == -1 && rhs.Coefficient != 0) {
			const double denRoot = -result.Constant / result.Coefficient;
			const double numRem = rhs.Coefficient * denRoot + rhs.Constant;
			if (numRem != 0) return Error(context.Offset - 1, "나눗셈의 결과가 일차식이 아닙니다.");

			result.Constant = result.Coefficient / rhs.Coefficient; // 일차식/일차식
			result.Coefficient = 0;
			continue;
		}

		if (result.Coefficient == 0 && rhs.Coefficient == 0) { // 상수항*상수항
			result.Constant *= std::pow(rhs.Constant, exp);
		} else if (result.Coefficient == 0) {
			result.Coefficient = result.Constant * rhs.Coefficient; // 상수항*일차식
			result.Constant *= std::pow(rhs.Constant, exp);
		} else { // 일차식*상수항
			result.Coefficient *= std::pow(rhs.Constant, exp);
			result.Constant *= std::pow(rhs.Constant, exp);
		}
	}

	return true;
}
bool ParseParen(ParseContext& context, Equation& result) {
	if (!RequireChar(context, '(')) return ParseTerm(context, result);
	else return ParseAddSub(context, result) && RequireChar(context, ')');
}
bool ParseTerm(ParseContext& context, Equation& result) {
	double sign = 1;
	while (RequireChar(context, '+') || RequireChar(context, '-')) {
		sign *= context.Required == '+' ? 1 : -1;
	}

	SkipEmptyChars(context);
	if (context.Offset == context.Equation.size()) return Error(context.Offset, "올바른 형식의 일차방정식이 아닙니다.");

	if (std::isdigit(context.Equation[context.Offset])) {
		auto end = std::find_if_not(context.Equation.begin() + context.Offset, context.Equation.end(), [](char c) { return std::isdigit(c); });
		if (end < context.Equation.end() && *end == '.') {
			end = std::find_if_not(end + 1, context.Equation.end(), [](char c) { return std::isdigit(c); });
		}

		const auto distance = static_cast<std::size_t>(std::distance(context.Equation.begin(), end));
		const double number = sign * std::stod(std::string(
			context.Equation.substr(context.Offset, distance - context.Offset)));
		context.Offset = distance;

		if (distance < context.Equation.size() && context.Equation[distance] == 'x') {
			result.Coefficient = number;
			++context.Offset;
		} else {
			result.Constant = number;
		}

		return true;
	} else if (context.Equation[context.Offset] == 'x') {
		result.Coefficient = sign;
		++context.Offset;

		return true;
	} else return Error(context.Offset, "올바른 형식의 일차방정식이 아닙니다.");
}

File: linearPharse.js
// origin https://gist.github.com/kmc7468/b07b82929030ca4a5ab8f77de67147d3 
// viral https://blog.naver.com/kmc7468/222494116955 

/* 
 * MIT License 
 * 
 * Copyright (c) 2021 kmc7468 
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy 
 * of this software and associated documentation files (the "Software"), to deal 
 * in the Software without restriction, including without limitation the rights 
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell 
 * copies of the Software, and to permit persons to whom the Software is 
 * furnished to do so, subject to the following conditions: 
 * 
 * The above copyright notice and this permission notice shall be included in all 
 * copies or substantial portions of the Software. 
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE 
 * SOFTWARE. 
 */ 

{ /// 

class Equation { 
	Coefficient = 0 
	Constant = 0 
	constructor( ... ar ) { Object .assign( this, ... ar ) } 
	} // -- Equation{} 

class ParseContext { 
	Equation 
	Offset = 0 
	
	Required 
	} // -- ParseContext{} 

let g_InputMessage 
	= getMessageReceiver 
		`x에 대한 일차방정식을 입력하세요: ( 취소 : ESC / Cancel ) ` 

let debugging // = true 

consoleTemplate `
일차방정식 계산 프로그램 
(C) 2021. kmc7468 All rights reserved. 

javascript by killofki 
주의사항: 괄호 앞/뒤에서 곱셈 연산자를 생략할 수 없습니다. 
주의사항: 변수는 x만 가능합니다. 

: error list : 
x*x=1
x*x=0
(x+1)*(x+x+2)=3

` 
	// -- consoleTemplate 

startPoint : while( true ) { 
	let equation = g_InputMessage() 
	consoleTemplate( equation ) 
	
	if ( isBreak( equation ) ) { 
		break startPoint 
		} // -- if isBreak 
	
	let result 
	try { result = ParseEquation( equation ) } 
	catch( err ) { 
		console .error( err ) 
		} // -- catch 
	
	if ( ! ( result instanceof Equation ) ) { 
		// maybe was error message 
		continue startPoint 
		} // -- if ! ParseEquation 
	
	let root = SolveEquation( result ) 
	; 
		root === Infinity 
			? 
				( consoleTemplate `해가 무수히 많습니다.` 
				, consoleTemplate `` 
				) 
		: Number .isNaN( root ) 
			? 
				( consoleTemplate `해가 없습니다.` 
				, consoleTemplate `` 
				) 
		
		: 
			( consoleTemplate `해는 x=${ root } 입니다.` 
			, consoleTemplate `` 
			) 
	
	continue startPoint 
	} // -- while true // -- startPoint 

// .. functions .. 

function isBreak( equation ) { 
	return ( 
		equation == null 
		|| equation .toUpperCase() === 'ESC' 
		|| equation === '취소' 
		|| equation .toLowerCase() === 'cancel' 
		) // -- return 
	} // -- isBreak() 

function ParseEquation( equation ) { 
	let context = new ParseContext 
	let requireOn = RequireCharOn( context ) 
	
	context .Equation = equation 
	
	let lhs = new Equation 
	let rhs = new Equation 
	
	switch( true ) { 
		
		case ! ParseAddSub( context, lhs ) : 
			return false 
		
		case ! requireOn `=` : 
			ErrorAt( context .Offset ) 
			throw `등호가 없습니다.` 
		
		case ! ParseAddSub( context, rhs ) : 
			return false 
		
		case context .Offset < equation .length : 
			ErrorAt( context .Offset ) 
			throw `올바른 형식의 일차방정식이 아닙니다.` 
		
		} // -- switch true 
	
	// equaling 
	let { Coefficient, Constant } = propMinus( lhs, rhs ) 
	Constant = -1 * Constant // rhs - lhs 
	let result = new Equation({ Coefficient, Constant }) 
	
	debugging 
		&& consoleTemplate `${ Coefficient }x = ${ Constant } // ParseEquation` 
	
	return result 
	} // -- ParseEquation() 

function SolveEquation( equation ) { 
	let { Coefficient, Constant } = equation 
	debugging 
		&& consoleTemplate `${ Coefficient }x = ${ Constant } // SolveEquation` 
	
	return ( 
		Coefficient === 0 
			? Constant === 0 ? Infinity : NaN 
			: Constant / Coefficient 
		) // -- return 
	} // -- SolveEquation() 

function SkipEmptyChars( context ) { 
	SkipEmptyChars 
		= matcherRail( /\s+/gy ) // lazy 
	
	return SkipEmptyChars( context ) // call 
	} // -- SkipEmptyChars() 

function RequireCharsOn( context ) { 
	let requireOn = RequireCharOn( context ) 
	
	return ( ... ar ) => { 
		let Cs = rawValue( ... ar ) 
		
		SkipEmptyChars( context ) 
		
		for ( let c of Cs ) { 
			if ( requireOn( c ) ) { 
				return true 
				} // -- if requireOn 
			} // -- for of Cs 
		return false 
		} // -- () // -- return 
	} // -- RequireCharsOn() 

function RequireCharOn( context ) { 
	return ( ... ar ) => { 
		let c = rawValue( ... ar ) 
		
		SkipEmptyChars( context ) 
		
		let { Offset, Equation } = context 
		
		if 
				(  Offset < Equation .length 
				&& Equation[ Offset ] === c 
				) { 
			context .Required = c 
			context .Offset += 1 
			
			return true 
			} // if < length 
		
		return false 
		} // -- () // -- return 
	} // -- RequireCharOn() 

function ErrorAt( offset ) { 
	let space = ' ' .repeat( offset ) 
	consoleTemplate `${ space }^` 
	} // -- ErrorAt() 

function ParseAddSub( context, result ) { 
	if ( ! ParseMulDiv( context, result ) ) { 
		return false 
		} // -- if ! ParseMulDiv 
	
	let requiresOn = RequireCharsOn( context ) 
	
	inSumSub : while ( requiresOn `+-` ) { 
		const sign = context .Required === '+' ? 1 : -1 
		let rhs = new Equation 
		if ( ! ParseMulDiv( context, rhs ) ) { 
			return false 
			} // -- if ! ParseMulDiv 
		
		let signRhs = multiplyToProp( rhs, sign ) // lazy 
		let { Coefficient, Constant } = propPlus( signRhs, result ) // calling 
		
		debugging 
			&& consoleTemplate 
				`${ Coefficient }x + ${ Constant } // ParseAddSub` 
		
		Object .assign( result, { Coefficient, Constant } ) 
		} // -- while RequireCharsOn +- // -- inSumSub 
	
	return true 
	} // -- ParseAddSub() 

function ParseMulDiv( context, result ) { 
	if ( ! ParseParen( context, result ) ) { 
		return false 
		} // -- if ! ParseParen 
	
	let requiresOn = RequireCharsOn( context ) 
	
	inMultiply : while ( requiresOn `*/` ) { 
		const mulTo 
			= context .Required === '*' ? multiplyToProp 
			: divideToProp 
		
		let rhs = new Equation 
		if ( ! ParseParen( context, rhs ) ) { 
			return false 
			} // -- if ! ParseParen 
		
		checkDivide({ mulTo, rhs, context }) 
		checkMulDiv({ result, rhs, context }) 
		
		let resultNoCoefficient 
			=  result .Coefficient === 0 
			&& mulTo === multiplyToProp 
		
		let [ singleOrder, constOrder ] 
			=  resultNoCoefficient ? [ rhs, result ] 
			: [ result, rhs ] 
				// rhs have no Coefficient when mulTo <- divideToProp 
				// by checkDivide 
		
		let { Coefficient, Constant } 
			= mulTo( singleOrder, constOrder .Constant ) 
		
		debugging 
			&& consoleTemplate 
				`${ Coefficient }x + ${ Constant } // continue < ParseMulDiv` 
		
		Object .assign( result, { Coefficient, Constant } ) 
		
		continue inMultiply 
		} // -- while requiresOn 
	
	return true 
	} // -- ParseMulDiv() 

function ParseParen( context, result ) { 
	let requireOn = RequireCharOn( context ) 
	
	let getSigned = RequireCharsOn( context ) `+-` // receive just single sign 
	let preSign = getSigned ? context .Required : '' 
	
	let parenReceived 
		= requireOn `(` 
			?  ParseAddSub( context, result ) 
			&& requireOn `)` 
		: ParseTerm( context, result ) 
	
	if ( preSign === '-' ) { 
		let { Coefficient, Constant } = multiplyToProp( result, -1 ) 
		debugging 
			&& consoleTemplate 
				`${ Coefficient }x + ${ Constant } // ParseAddSub` 
		
		Object .assign( result, { Coefficient, Constant } ) 
		} // -- preOperator 
	
	return parenReceived 
	} // -- ParseParen() 

function checkDivide({ mulTo, rhs, context }) { 
	if ( mulTo !== divideToProp ) 
		{ return } 
	
	if ( rhs .Coefficient !== 0 ) { 
		ErrorAt( context .Offset - 1 ) 
		throw `일차식으로 나눌 수 없습니다.`  
		} // -- if !== 0 
		
	if ( rhs .Constant === 0 ) { 
		ErrorAt( context .Offset - 1 ) 
		throw `0으로 나눌 수 없습니다.` 
		} // -- if === 0 
	} // -- checkDivide() 

function checkMulDiv({ rhs, result, context }) { 
	let isNotSingleMul 
		=  rhs .Coefficient != 0 
		&& result .Coefficient != 0 
	if ( isNotSingleMul ) { 
		ErrorAt( context .Offset - 1 ) 
		throw `곱셈/나눗셈의 결과가 일차식이 아닙니다.` 
		} // -- if != 0 
	} // -- checkMulDiv() 

function ParseTerm( context, result ) { 
	let { X, number } = matchingNumberXAt( context ) 
	
	if ( X ) { 
		debugging 
			&& consoleTemplate `${ number }x // ParseTerm` 
		
		result .Coefficient = number 
		
		return true 
		} // -- if X 
	
	debugging 
		&& consoleTemplate `${ number } // ParseTerm` 
	
	result .Constant = number 
	
	return true 
	} // -- ParseTerm() 

function matchingNumberXAt( context ) { 
	let { Offset } = context 
	
	let matchingSign 
		= matcherRail( /\s*(?<signChars>[+-]+)/gy ) 
	let matchingNumberString 
		= matcherRail( /\s*(?<numberString>\d+(\.\d+)?)/gy ) 
	let matchingX 
		= matcherRail( /\s*(?<X>x)/gy ) 
	
	let { signChars = '' } = matchingSign( context ) ?.groups ?? {} 
	let { numberString } = matchingNumberString( context ) ?.groups ?? {} 
	let { X } = matchingX( context ) ?. groups ?? {} 
	
	let foundNumberX = numberString ?? X 
	if ( ! foundNumberX ) { 
		ErrorAt( Offset ) 
		throw `올바른 형식의 일차방정식이 아닙니다.` 
		} // -- if ! foundNumberX 
	
	numberString ??= '1' 
	
	let sign = 1 
	for ( let signChar of signChars ) { 
		sign *= `${ signChar }1` 
		} // -- for of signs 
	
	let number = sign * numberString 
	
	return { X, number } 
	} // -- matchingNumberXAt() 

function matcherRail( matcher ) { // must /()/gy 
	let railBody = 
		function getNext( context ) { 
			let { Offset, Equation } = context 
			
			matcher .lastIndex = Offset 
			let matched = matcher .exec( Equation ) 
			
			if ( matched ?.index === Offset ) { 
				context .Offset = matcher .lastIndex 
				
				return matched 
				} // -- if matched 
			} // -- getNext() 
	
	return railBody 
	} // -- matcherRail() 

function propPlus( left, right ) { 
	let get = ( t, prop ) => left[ prop ] + right[ prop ] 
	
	return new Proxy( {}, { get } ) 
	} // -- propPlus() 

function propMinus( left, right ) { 
	let get = ( t, prop ) => left[ prop ] - right[ prop ] 
	
	return new Proxy( {}, { get } ) 
	} // -- propMinus() 

function multiplyToProp( obj, multTime ) { 
	let get = ( t, prop ) => obj[ prop ] * multTime 
	
	return new Proxy( {}, { get } ) 
	} // -- multiplyToProp() 

function divideToProp( obj, multTime ) { 
	let get = ( t, prop ) => obj[ prop ] / multTime 
	
	return new Proxy( {}, { get } ) 
	} // -- divideToProp() 

function getMessageReceiver( ... ar ) { 
	let order = rawValue( ... ar ) 
	return q => { 
		consoleTemplate( order ) 
		
		return prompt( order ) 
		} // -- () // -- return 
	} // -- getMessageReceiver()  

function consoleTemplate( ... ar ) { 
	let t = rawValue( ... ar ) 
	
	console .log( t ) 
	} // -- consoleTemplate() 

function rawValue( ... ar ) { 
	let [ rawo ] = ar 
	
	return rawo ?.raw ? String .raw( ... ar ) 
		: rawo 
	} // -- rawValue() 

} /// 
